aws-sdk-resources/lib/aws-sdk-resources/services/s3/file_downloader.rb

```diff
@@ -0,0 +1,175 @@
+        part_size = (file_size.to_f / count.to_f).ceil
+        if chunk_size < part_size
+          get_range(construct_chunks(file_size))
+        else
+          get_part(count)
+        end
+      end
+
+      def construct_chunks(file_size)
+        offset = 0
+        default_chunk_size = compute_chunk(file_size)
+        chunks = []
+        while offset <= file_size
+          progress = offset + default_chunk_size
+          if progress < file_size
+            chunks << "bytes=#{offset}-#{progress}"
+          else
+            chunks << "bytes=#{offset}-#{file_size}"
+          end
+          offset = progress + 1
+        end
+        chunks
+      end
+
+      def compute_chunk(file_size)
+        if @chunk_size && @chunk_size > file_size
+          raise ArgumentError, ":chunk_size shouldn't exceed total file size."
+        else
+          default_chunk_size = @chunk_size || [(file_size.to_f / MAX_PARTS).ceil, MIN_CHUNK_SIZE].max.to_i
+        end
+      end
+
+      def sort_files(files)
+        # sort file by start range count or part number
+        files.sort do |a, b|
+          a[/([^\=]+)$/].split('-')[0].to_i <=> b[/([^\=]+)$/].split('-')[0].to_i
+        end
+      end
+
+      def concatenate_files(fileparts)
+        begin
+          concatenate_parts(fileparts)
+        ensure
+          fileparts.each do |part|
+            File.unlink(part) if File.exists?(part)
+          end
+        end
+      end
+
+      def concatenate_parts(fileparts)
+        File.open(@path, 'w', encoding: 'UTF-8') do |output_path|
+          sort_files(fileparts).each do |part|
+            File.open(part, 'r', encoding: 'UTF-8') { |f| output_path.write(f.read) }
+          end
+        end
+      end
+
+      def file_batches(chunks, mode)
+        batches = []
+        dir = Dir.tmpdir
+        chunks = (1..chunks) if mode.eql? 'part_number'
+        chunks.each_slice(@thread_count) do |slice|
+          batches << map_files(slice, dir, mode)
+        end
+        batches
+      end
+
+      def map_files(slice, dir, mode)
+        case mode
+        when 'range'
+          slice.inject({}) {|h, chunk| h[chunk] = File.join(dir, chunk); h}
+        when 'part_number'
+          slice.inject({}) {|h, part| h[part] = File.join(dir, "part_number=#{part}"); h}
+        end
+      end
+
+      def get_range(chunks)
+        thread_batches(chunks, 'range')
+      end
+
+      def get_part(parts)
+        thread_batches(parts, 'part_number')
+      end
+
+      def thread_batches(chunks, param)
+        batches = file_batches(chunks, param)
+        batches.each do |batch|
+          threads = []
+          batch.each do |chunk, file|
+            threads << Thread.new do
+              resp = @client.get_object(
+                :bucket => @bucket,
+                :key => @key,
+                param.to_sym => chunk
+              )
+              File.open(file, 'w', encoding: 'UTF-8') {|f| f.write(resp.body.read)}
```

Comment:
Here you can use the `:response_target` option with a path, which will automatically download the file in chunks, and also use the (correct) binary format.

```rb
@client.get_object(
  :bucket => @bucket,
  :key => @key,
  param.to_sym => chunk,
  :response_target => file
)
```